### PR TITLE
feat(SelectControl): add children prop

### PR DESCRIPTION
**Pull request by frosso**
## Description

It looks like the `select-control` component accepts an `options` prop, which is just an array of options that are mapped to `<option />` elements.
`<optgroup />`s are a native part of JSX. However, they cannot be added through the current `options` prop.

I understand that `options` is provided as an abstraction, but I think that rather than changing the `options` prop, we can leverage the native `children` prop present on all the JSX elements and obtain the same result.
We could still leverage the `options` prop to add `<optgroup />` elements, but where's the fun in that? It seems to me that by leveraging the `children` prop we can have a more semantical component. And with that we can just leverage `<optgroup />`s natively!

I kept the `options` prop, but maybe it can deprecated in favor of `children`?

## How has this been tested?

I tested this in Storybook. I also added some unit tests for the component (there weren't any before).

## Screenshots 
https://d.pr/i/JvrB57

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] I've tested my changes with keyboard and screen readers. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

File: packages/components/src/select-control/test/select-control.test.js [hunk 1]
```diff
@@ -0,0 +1,82 @@
+/**
```
Comment (talldan) [inline packages/components/src/select-control/test/select-control.test.js]: A convention in the codebase is that test file name should try to map to the filename that's being tested in the parent folder, so here it should be called `index.js` (and it doesn't need the `test.js` postfix).

**Review by talldan — COMMENTED** [review]
Thanks for your contribution @frosso.

This sounds like a reasonable change, and probably how many would expect the component to work. From memory, I think there are other components that work like this too (maybe `DropdownMenu`?)

I think as part of this the README.md would need to be updated. Some of the docs may (or may not) need to be generated after modifying that file (`npm run docs:build` and then also commit and push any generated changes).

File: packages/components/src/select-control/test/select-control.test.js (hunk 1 above)
Comment (frosso) [reply packages/components/src/select-control/test/select-control.test.js]: Thanks - done in 6c4e3a219e

**Review by frosso — COMMENTED** [review]
Thanks for taking the time to review @talldan !
I made an update to the `README.md` in 18872a4300. Running `npm run docs:build` didn't seem to create additional changes.
Let me know if there's anything else!

**Review by ajlende — APPROVED** [review]
I resolved the conflicts with trunk and everything works as expected. Looks like @talldan's comments have all been resolved too, so I think this looks good to go :+1:

File: packages/components/src/select-control/README.md [hunk 2]
```diff
@@ -174,6 +199,13 @@ An array of objects containing the following properties:
 -   Type: `Array`
 -   Required: No
 
+#### children
+
+An alternative to the `options` prop.  
+Use the `children` prop to have more control on the style of the items being rendered, like `optgroup`s or `options` and possibly avoid re-rendering due to the reference update on the `options` prop.
+- Type: `Element`
```
Comment (ciampo) [inline packages/components/src/select-control/README.md:206]: `ReactNode` would be a better choice for describing `children` 's type

**Review by ciampo — COMMENTED** [review]
Hey @frosso , it looks like this change introduced some typescript errors in the package, since the compiler now thinks that the `children` prop is mandatory on `SelectControl` (for example, [here](https://github.com/WordPress/gutenberg/blob/1e56c54f832b8c7284f1516d81f6511927f739ce/packages/components/src/color-picker/component.tsx#L104-L112)) — would you be able to work on a fix to mark `children` as optional?

<details>
<summary>Something like this should be enough</summary>

```diff
diff --git a/packages/components/src/select-control/index.tsx b/packages/components/src/select-control/index.tsx
index 7f68f56721..ec2afed60d 100644
--- a/packages/components/src/select-control/index.tsx
+++ b/packages/components/src/select-control/index.tsx
@@ -4,7 +4,7 @@
 import { isEmpty, noop } from 'lodash';
 import classNames from 'classnames';
 // eslint-disable-next-line no-restricted-imports
-import type { ChangeEvent, FocusEvent, Ref } from 'react';
+import type { ChangeEvent, FocusEvent, ReactNode, Ref } from 'react';
 
 /**
  * WordPress dependencies
@@ -31,7 +31,7 @@ function useUniqueId( idProp?: string ) {
 }
 
 export interface SelectControlProps
-	extends Omit< InputBaseProps, 'isFocused' > {
+	extends Omit< InputBaseProps, 'children' | 'isFocused' > {
 	help?: string;
 	hideLabelFromVision?: boolean;
 	multiple?: boolean;
@@ -50,6 +50,7 @@ export interface SelectControlProps
 	size?: Size;
 	value?: string | string[];
 	labelPosition?: LabelPosition;
+	children?: ReactNode;
 }
 
 function SelectControl(
```
</details>

Also, when making changes on the `@wordpress/components` package, we usually add an entry to the [CHANGELOG](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CHANGELOG.md). You can find more information regarding best practices in the [contributing guidelines](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md).

Finally, feel free to ping myself and/or @mirka in any future PR concerting `@wordpress/components`. Thank you!

**Comment by ciampo** [comment]
Opened #37872 with the fix suggested in the previous comment

**Pull request merged**